packages/core/src/utils/operators.ts

```diff
@@ -0,0 +1,13 @@
+import type { BaseEvent } from '../agent/Events'
+import type { BaseRecord } from '../storage/BaseRecord'
+
+import { filter, pipe } from 'rxjs'
+
+/**
+ *
+ * @param type Record type (extends BaseRecord type)
+ * @returns UnaryFunction
+ */
+export function filterByRecordType<T extends BaseRecord<any, any, any>>(type: string) {
+  return pipe(filter((e: BaseEvent) => (e.payload as Record<string, T>).record.type === type))
```

Comment:
If I call it like this we have full typing info:

```ts
agent.events.observable<ConnectionStateChangedEvent>(ConnectionEventTypes.StateChanged).pipe(
        // connectionRecord is of type ConnectionRecord here because we passed the `ConnectionStateChangedEvent`
	// generic
	tap(event => event.payload.connectionRecord)
)
```

So what if we make the method so the event must have a base record param:

```ts
export function filterByRecordType<R extends BaseRecord>(recordClass: Constructor<R> & { type: string }) {
  return <Event extends { payload: { record: BaseRecord } }>(source: Observable<Event>) =>
    source.pipe(
      filter((e: Event): e is Event & { payload: { record: R } } => e.payload.record.type === recordClass.type)
    )
}
```

And then use it like:

```ts
agent.events.observable<RecordDeletedEvent>(RepositoryEventTypes.RecordDeleted).pipe(
  filterByRecordType(ConnectionRecord),
  // record is fully typed now
  tap((e) => e.payload.record.assertReady())
)
```